File: backend-springboot/src/main/java/com/greenalgeria/zone/application/command/RegisterVolunteerHandler.java
package com.greenalgeria.zone.application.command;

import com.greenalgeria.shared.cqrs.CommandHandler;
import com.greenalgeria.shared.exception.NotFoundException;
import com.greenalgeria.zone.application.*;
import com.greenalgeria.zone.domain.ZoneRepository;
import org.springframework.stereotype.Component;
import org.springframework.transaction.annotation.Transactional;

@Component
@Transactional
public class RegisterVolunteerHandler implements CommandHandler<RegisterVolunteerCommand, Void> {

    private final ZoneRepository zoneRepository;

    public RegisterVolunteerHandler(ZoneRepository zoneRepository) {
        this.zoneRepository = zoneRepository;
    }

    public Void handle(RegisterVolunteerCommand command) {
        var zone = zoneRepository.findById(command.id()).orElseThrow(() -> new NotFoundException("Zone not found"));
        zone.incrementVolunteers();
        zoneRepository.save(zone);
        return null;
    }

    @Override
    public Class<RegisterVolunteerCommand> supportedCommand() {
        return RegisterVolunteerCommand.class;
    }
}

File: backend-nestjs/src/modules/zones/application/commands/register-volunteer.handler.ts
import { CommandHandler, ICommandHandler } from '@nestjs/cqrs';
import { Inject, NotFoundException } from '@nestjs/common';
import type { Cache } from 'cache-manager';
import { CACHE_MANAGER } from '@nestjs/cache-manager';
import { RegisterVolunteerCommand } from './register-volunteer.command';
import { ZoneRepository } from '../../infrastructure/zone.repository';

@CommandHandler(RegisterVolunteerCommand)
export class RegisterVolunteerHandler implements ICommandHandler<RegisterVolunteerCommand> {
  constructor(
    private readonly repository: ZoneRepository,
    @Inject(CACHE_MANAGER) private readonly cache: Cache,
  ) {}

  async execute(command: RegisterVolunteerCommand): Promise<void> {
    const zone = await this.repository.findById(command.zoneId);
    if (!zone) throw new NotFoundException(`Zone ${command.zoneId} not found`);

    zone.registerVolunteer();
    await this.repository.save(zone);
    await this.cache.del('zones:all');
    await this.cache.del(`zone:${command.zoneId}`);
  }
}
